### PR TITLE
capture duration and label with error source

### PR DESCRIPTION
**Pull request by scottlepp**
* capture duration as a prometheus metric
* add error source label
* refactor a bit.  the QueryDB function had a lot of params.  try to reduce the arity

### CODE REVIEW

File: query.go [hunk 1]
```diff
@@ -203,3 +246,29 @@ func applyHeaders(query *Query, headers http.Header) *Query {
 
 	return query
 }
+
+// sanitizeLabelName removes all invalid chars from the label name.
+// If the label name is empty or contains only invalid chars, it
+// will return an error.
```
Comment (cletter7) [inline query.go:252]: [nit] the function doesn't return an error, update the comment?

File: query.go [hunk 2]
```diff
@@ -51,16 +60,47 @@ func GetQuery(query backend.DataQuery, headers http.Header, setHeaders bool) (*Q
...
+		DB:         db,
+		DSName:     dsName,
+		converters: converters,
+		fillMode:   fillMode,
+	}
+}
+
+// Run sends the query to the connection and converts the rows to a dataframe.
+func (q *DBQuery) Run(ctx context.Context, query *Query, args ...interface{}) (data.Frames, error) {
 	// Query the rows from the database
```
Comment (cletter7) [inline query.go:86]: This comment should be before line 96 it seems. Or maybe not needed at all.

File: query.go [hunk 3]
```diff
@@ -203,3 +246,29 @@ func applyHeaders(query *Query, headers http.Header) *Query {
 
 	return query
 }
+
+// sanitizeLabelName removes all invalid chars from the label name.
+// If the label name is empty or contains only invalid chars, it
+// will return an error.
+func sanitizeLabelName(name string) (string, bool) {
```
Comment (cletter7) [inline query.go:253]: Should we add test for this function?

File: query.go [hunk 4]
```diff
@@ -51,16 +60,47 @@ func GetQuery(query backend.DataQuery, headers http.Header, setHeaders bool) (*Q
...
+	}()
+
+	rows, err := q.DB.QueryContext(ctx, query.RawSQL, args...)
 	if err != nil {
 		errType := ErrorQuery
 		if errors.Is(err, context.Canceled) {
 			errType = context.Canceled
 		}
 		err := DownstreamError(fmt.Errorf("%w: %s", errType, err.Error()))
+		errorSource = string(backend.ErrorSourceDownstream)
```
Comment (cletter7) [inline query.go:103]: Don't understand why do we need `errorSource` separately? Can't we get the error source from the error itself?

File: query.go (hunk 3 above)
Comment (scottlepp) [reply query.go:253]: this was copied from https://github.com/grafana/grafana/blob/main/pkg/infra/metrics/metricutil/utils.go#L14

Comment (cletter7) [reply query.go:253]: Should we then copy test also?
https://github.com/grafana/grafana/blob/main/pkg/infra/metrics/metricutil/utils_test.go#L11-L33

File: query.go (hunk 4 above)
Comment (scottlepp) [reply query.go]: updated

**Review by cletter7 — APPROVED** [review]

**Pull request merged**